todo_app/app.py

```diff
@@ -1,92 +1,223 @@
+from flask_login import login_required
+from flask_login import LoginManager
+from flask_login import login_user
+from flask_login import current_user
 import os
+import requests
+import json
+from oauthlib.oauth2 import WebApplicationClient
 
-def create_app(db_name = ""):
+
+
+def create_app():
    app = Flask(__name__)
    item_view_model = None
 
+   disable_login = os.getenv('FLASK_SKIP_LOGIN')
+   if disable_login:
+      print("Switching off authentication in the app config for testing purposes")
+      app.config['LOGIN_DISABLED'] = True
    mongo_srv = os.getenv('MONGO_SRV')
-   if db_name == "":
-      db_name = os.getenv('MONGO_DB')
+   db_name = os.getenv('MONGO_DB')
    mongo_user = os.getenv('MONGO_USER')
    mongo_pwd = os.getenv('MONGO_PWD')
    mongo_connection = os.getenv('MONGO_CONNECTION')
 
    mongo_client = ToDoMongoClient(mongo_user, mongo_pwd, mongo_srv, db_name, mongo_connection)
 
+   oauth_client_id = os.getenv('OAUTH_CLIENT_ID')
+   oauth_secret_id = os.getenv('OAUTH_CLIENT_SECRET')
+
+   login_manager = LoginManager() 
+   login_manager.init_app(app)
+   client = WebApplicationClient(oauth_client_id)
+
+   os.environ['OAUTHLIB_INSECURE_TRANSPORT'] = '1'
+   app.secret_key = oauth_secret_id
+
+   writer_role = "Writer"
+   reader_role = "Reader"
+   admin_role = "Admin"
+
+
    #  All the routes and setup code etc
    @app.route('/')
+   @login_required
    def index():
+      print(f"Current user is {current_user}")
+      
       show_all = request.args.get('show_all')
       
       show_all_bool = show_all == "yes"
-      print(f"Show all value is {show_all_bool}")
       tasks = mongo_client.get_all_tasks()
       item_view_model = ViewModel(tasks)
+      # We want to only show certain buttons and links to writers or admins UNLESS login is disabled (ie. we are running tests)
+      writer_bool = (disable_login or current_user.role == writer_role)
+      admin_bool = (disable_login or current_user.role == admin_role)
       return render_template('index.html', title='To Do App',
-         view_model=item_view_model, show_all=show_all_bool)
-
+         view_model=item_view_model, show_all=show_all_bool, writer=writer_bool, admin=admin_bool)
 
    @app.route('/items/add', methods = ['POST'])
+   @login_required
    def add_item():
-      form_data = request.form
-      task_title = form_data["title"]
-      task_desc = form_data["description"]
-      task_due_date = form_data["duedate"]
-      
-      mongo_client.add_task(task_title, task_desc, task_due_date) 
-      
+      if (disable_login or current_user.role == writer_role):
+         form_data = request.form
+         task_title = form_data["title"]
+         task_desc = form_data["description"]
+         task_due_date = form_data["duedate"]
+         
+         mongo_client.add_task(task_title, task_desc, task_due_date) 
+      else:
+         print("Reader Role is not allowed to add a task")
```

Comment:
Perfectly good solution here - but if this was a live app we'd probably want to make it clear to the user that their action hadn't had any effect so that they understand that their request has been denied